L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc

```diff
@@ -0,0 +1,112 @@
+#include "L1Trigger/Phase2L1ParticleFlow/interface/L1NNTauProducer.hh"
+#include <TLorentzVector.h>
+#include <cmath>
+
+L1NNTauProducer::L1NNTauProducer(const edm::ParameterSet& cfg)
+    : fSeedPt_(cfg.getParameter<double>("seedpt")),
+      fConeSize_(cfg.getParameter<double>("conesize")),
+      fTauSize_(cfg.getParameter<double>("tausize")),
+      fMaxTaus_(cfg.getParameter<int>("maxtaus")),
+      fNParticles_(cfg.getParameter<int>("nparticles")),
+      fL1PFToken_(consumes<vector<l1t::PFCandidate> >(cfg.getParameter<edm::InputTag>("L1PFObjects"))) {
+  std::string lNNFile = cfg.getParameter<std::string>("NNFileName");  //,"L1Trigger/Phase2L1Taus/data/tau_3layer.pb");
+  fTauNNId_ = std::make_unique<TauNNId>();
+  if (lNNFile.find("v0") == std::string::npos)
+    fTauNNId_->initialize("input_1:0", lNNFile, fNParticles_);
+  else if (lNNFile.find("v0") != std::string::npos)
+    fTauNNId_->initialize("dense_1_input:0", lNNFile, fNParticles_);
+  produces<l1t::PFTauCollection>("L1PFTausNN");
+}
+
+void L1NNTauProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::Handle<l1t::PFCandidateCollection> l1PFCandidates;
+  iEvent.getByToken(fL1PFToken_, l1PFCandidates);
+
+  l1t::PFCandidateCollection pfChargedHadrons;
+  l1t::PFCandidateCollection pfChargedHadrons_sort;
+  l1t::PFCandidateCollection pfChargedHadrons_seeds;
+  for (const auto& l1PFCand : *l1PFCandidates)
+    if ((l1PFCand.id() == l1t::PFCandidate::ChargedHadron || l1PFCand.id() == l1t::PFCandidate::Electron) &&
+        std::abs(l1PFCand.eta()) < 2.5)
+      pfChargedHadrons_sort.push_back(l1PFCand);
+  std::sort(pfChargedHadrons_sort.begin(), pfChargedHadrons_sort.end(), [](l1t::PFCandidate i, l1t::PFCandidate j) {
+    return (i.pt() > j.pt());
+  });
+  auto lTaus = std::make_unique<l1t::PFTauCollection>();
+  if (pfChargedHadrons_sort.empty()) {
+    if (lTaus->empty()) {
+      PFTau dummy;
+      lTaus->push_back(dummy);
+    }
+    iEvent.put(std::move(lTaus), "L1PFTausNN");
+    return;
+  }
+  pfChargedHadrons_seeds.push_back(pfChargedHadrons_sort[0]);
```

Comment:
there is a lot of copying happening here and below. can some of the copies be avoided?

Reply:
I would really prefer to address this comment and possibly simplify it once the code is in and a proper unit tests/RelVals are available.  I would really like to avoid changing this part of the code here as there a bug might get introduced unintentionally.

Reply:
If there are concerns about bugs, unit tests should be part of this PR. Comparison tests after the code is merged show consistency, not correctness.

Reply:
The code "as-is" is validated and shows the expected physics performance of the reconstructed.  So any changes in comparison tests produced by re-arranging the code would be a sign of unintentional change.  Unit tests cannot be part of this PR, as they will require ntuples which are not here yet.

Reply:
I still don't really understand how making `pfChargedHadrons_sort` and `pfChargedHadrons_seeds` vectors of pointers to candidates (rather than copying the candidates each time) is prone to introduce bugs.

Reply:
would still prefer to see this addressed in this PR.

Reply:
hi @kpedro88 
addressing this now.